src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs

```diff
@@ -219,48 +218,34 @@ private unsafe bool WriteSubmissions()
                     }
                 }
                 _iovsUsed = iovIndex;
-                _sqesQueued = (uint)(ring.SubmissionQueueSize - sqesAvailable);
 
                 bool operationsRemaining = (_newOperations.Count - _newOperationsQueued) > 0;
                 return operationsRemaining;
             }
 
-            public unsafe void SubmitAndWait(bool mayWait)
+            public unsafe bool SubmitAndWait(bool mayWait)
             {
-                try
-                {
-                    bool operationsRemaining = WriteSubmissions();
-
-                    // We can't wait if there are more submissions to be sent,
-                    // or the event loop wants to do something.
-                    bool waitForCompletion = !operationsRemaining && mayWait;
+                bool operationsRemaining = WriteSubmissions();
 
-                    // io_uring_enter
-                    _ring!.Submit();
-                    uint submitted = _ring!.Flush((uint)_sqesQueued, minComplete: waitForCompletion ? 1U : 0);
-                    _sqesQueued -= submitted;
+                // We can't wait if there are more submissions to be sent,
+                // or the event loop wants to do something.
+                bool waitForCompletion = !operationsRemaining && mayWait;
 
-                    if (_sqesQueued == 0) // likely case: all sqes were queued
-                    {
-                        _iovsUsed = 0;
-                        _newOperations.RemoveRange(0, _newOperationsQueued);
-                        _newOperationsQueued = 0;
-                    }
-                    else
+                // io_uring_enter
+                SubmitResult result;
+                do
+                {
+                    result = _ring!.SubmitAndWait(minComplete: waitForCompletion ? 1U : 0, out _);
+                    if (result == SubmitResult.AwaitCompletions)
                     {
-                        // We were not able to submit all requests.
-
-                        // TODO: This seems similar to EAGAIN, not enough resources?
-                        // Or does it happen in other cases?
-                        // Is there a semantical difference between 0 and EAGAIN;
-                        // could submitted be less than _seqsQueued if there is an issue with
-                        // the sqe at submitted + 1?
-                        // TODO: detect if we're not making any more progress.
+                        return false;
                     }
-                }
-                catch (ErrnoException ex) when (ex.Errno == EBUSY || // The application needs to read completions.
-                                                ex.Errno == EAGAIN)  // The kernel doesn't have enough resources.
-                { }
+                } while (result == SubmitResult.SubmittedPartially); // TODO: detect if we're not making any more progress.
```

Comment:
I removed the while loop in https://github.com/tmds/Tmds.LinuxAsync/pull/23.
Does the new API require this to be a while loop?

Reply:
Can you remove the `while` loop
and replace the `_sqesQueued == 0` with `result = SubmittedSuccessfully`?

Reply:
> I removed the while loop

I missed that... sorry.

> Does the new API require this to be a while loop?

No, it's just a possibility that not all submissions were consumed by the kernel. We can chose to ignore that and work our way around it without looping.
If we just remove the loop, the next `WriteSubmissions` has less `sqesAvailable` available for submissions. The next `ring.SubmitAndWait` will then likely submit all pending entries and everything is back to "normal".

> Can you remove the while loop and replace the _sqesQueued == 0 with result = SubmittedSuccessfully?

Not sure I understand...

Reply:
> If we just remove the loop, the next WriteSubmissions has less sqesAvailable available for submissions. 

Yes, this falls back to the outer loop to submit them later. Can you remove the `while` loop?